haskell.compiler.ghc91{0,2}: provide LLVM compatible assembler

sternenseemann opened the pull request:
Verified that there are no hash changes (eval platform x86_64-linux) due to `getToolExe` in:

- haskell.compiler
- pkgsCross.armv7l-hf-multiplatform.haskell.compiler
- pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler


Tested `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc9102` with `LLVMAS`.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

sternenseemann committed:
haskell.compiler.*: add helper for calculating LLVM tools' paths

Verified that there are no hash changes (eval platform x86_64-linux) in:

- haskell.compiler
- pkgsCross.armv7l-hf-multiplatform.haskell.compiler
- pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler

sternenseemann committed:
haskell.compiler.ghc91{0,2}: provide LLVM compatible assembler

If we don't explicitly pass something as LLVMAS, GHC >= 9.10 will try to
invoke "clang" at runtime which fails.

Upstream change: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12005

I've tested pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc9102
which was reported to be failing here:
https://github.com/NixOS/nixpkgs/pull/440271#issuecomment-3262885486

emilazy reviewed (approved):
Looks correct to me. Haven’t tested, but it can’t make anything worse for Hydra‐supported platforms and is already broken for LLVM‐using ones, so it seems harmless to land preemptively.

It‘d be nice to land this on `master`, as it will make it easier to test https://github.com/NixOS/nixpkgs/pull/440774 with these versions. (Unless we’re sending that one to `haskell-updates` too.)

sternenseemann commented:
Yeah, I'll rebase it tomorrow.

sternenseemann commented:
> Now conflicted with 8.10.7.nix.

😢

Rebased.

emilazy commented:
I guess `forceLlvmCodegenBackend` may need updating? Or just dropping – it looks like it is only used for 9.6.3, which seems like it won’t be long for this world anyway.

sternenseemann committed:
{forceLlvmCodegenBackend,ghcWithPackages}: update clang -fllvm logic

- The LLVM backend always needs an LLVM specific assembler, i.e. clang
  that ideally matches the version of LLVM actually used for codegen.

- The LLVM backend on Darwin requires some version of clang to be
  available. In light of LLVMAS, using a matching version seems to be
  best though this does risk messing with clang used for compiling C
  in the derivation.

  This mess can be avoided by compiling GHC with useLLVM = true which
  sets absolute paths in GHC's settings file. Due to closure size
  constraints, we can't really do that if NCG is available.

emilazy reviewed (approved):
Still haven’t tested, but looks sensible.
